bundles/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/handler/AVMFritzBaseThingHandler.java

```diff
@@ -78,15 +94,202 @@ public void initialize() {
+                                    ZoneId.systemDefault())));
+                }
+                if (TEMP_FRITZ_UNDEFINED.equals(device.getHkr().getNextchange().getTchange())) {
+                    updateThingChannelState(CHANNEL_NEXTTEMP, UnDefType.UNDEF);
+                } else {
+                    updateThingChannelState(CHANNEL_NEXTTEMP, new QuantityType<>(
+                            toCelsius(device.getHkr().getNextchange().getTchange()), SIUnits.CELSIUS));
+                }
+            }
+            if (device.getHkr().getBattery() == null) {
+                updateThingChannelState(CHANNEL_BATTERY, UnDefType.UNDEF);
+            } else {
+                updateThingChannelState(CHANNEL_BATTERY, new DecimalType(device.getHkr().getBattery()));
+            }
+            if (device.getHkr().getBatterylow() == null) {
+                updateThingChannelState(CHANNEL_BATTERY_LOW, UnDefType.UNDEF);
+            } else {
+                updateThingChannelState(CHANNEL_BATTERY_LOW,
+                        BATTERY_ON.equals(device.getHkr().getBatterylow()) ? OnOffType.ON : OnOffType.OFF);
+            }
+        }
+        if (device instanceof DeviceModel && device.isAlarmSensor() && ((DeviceModel) device).getAlert() != null) {
+            updateThingChannelState(CHANNEL_CONTACT_STATE,
+                    AlertModel.ON.equals(((DeviceModel) device).getAlert().getState()) ? OpenClosedType.OPEN
+                            : OpenClosedType.CLOSED);
+        }
+        if (device instanceof DeviceModel && device.isButton() && ((DeviceModel) device).getButton() != null) {
+            if (((DeviceModel) device).getButton().getLastpressedtimestamp() == 0) {
+                updateThingChannelState(CHANNEL_LAST_CHANGE, UnDefType.UNDEF);
+            } else {
+                ZoneId zoneId = ZoneId.systemDefault();
+                ZonedDateTime timestamp = ZonedDateTime.ofInstant(
+                        Instant.ofEpochSecond(((DeviceModel) device).getButton().getLastpressedtimestamp()), zoneId);
+                Instant then = timestamp.toInstant();
+                ZonedDateTime now = ZonedDateTime.now(zoneId);
+                Instant someSecondsEarlier = now.minusSeconds(15).toInstant();
+                if (then.isAfter(someSecondsEarlier) && then.isBefore(now.toInstant())) {
+                    triggerThingChannel(CHANNEL_PRESS, CommonTriggerEvents.PRESSED);
+                }
+                updateThingChannelState(CHANNEL_LAST_CHANGE, new DateTimeType(timestamp));
+            }
+        }
+    }
+
+    /**
+     * Updates thing properties.
+     *
+     * @param device the {@link AVMFritzBaseModel}
+     */
+    private void updateProperties(AVMFritzBaseModel device) {
+        Map<String, String> editProperties = editProperties();
+        editProperties.put(Thing.PROPERTY_FIRMWARE_VERSION, device.getFirmwareVersion());
+        if (device instanceof GroupModel && ((GroupModel) device).getGroupinfo() != null) {
```

Comment:
`((GroupModel) device).getGroupinfo()` should be cached to a local variable

Reply:
Yes, I am aware of all those minor improvements. To be honest, I am not very happy with the current code. My plan is to refactor the binding step by step.